app/file_helper.go

```diff
@@ -0,0 +1,209 @@
+func (a *App) filterInaccessibleFiles(fileList *model.FileInfoList, options filterFileOptions) *model.AppError {
+	if fileList == nil || fileList.FileInfos == nil || len(fileList.FileInfos) == 0 {
+		return nil
+	}
+
+	lastAccessibleFileTime, appErr := a.GetLastAccessibleFileTime()
+	if appErr != nil {
+		return model.NewAppError("filterInaccessibleFiles", "app.last_accessible_file.app_error", nil, appErr.Error(), http.StatusInternalServerError)
+	}
+	if lastAccessibleFileTime == 0 {
+		// No need to filter, all files are accessible
+		return nil
+	}
+
+	if len(fileList.FileInfos) == len(fileList.Order) && options.assumeSortedCreatedAt {
+		lenFiles := len(fileList.FileInfos)
+		getCreateAt := func(i int) int64 { return fileList.FileInfos[fileList.Order[i]].CreateAt }
+
+		bounds := getTimeSortedPostAccessibleBounds(lastAccessibleFileTime, lenFiles, getCreateAt)
+
+		if bounds.allAccessible(lenFiles) {
+			return nil
+		}
+		if bounds.noAccessible() {
+			if lenFiles > 0 {
+				firstFileCreatedAt := fileList.FileInfos[fileList.Order[0]].CreateAt
+				lastFileCreatedAt := fileList.FileInfos[fileList.Order[lenFiles-1]].CreateAt
+				fileList.FirstInaccessibleFileTime = max(firstFileCreatedAt, lastFileCreatedAt)
+			}
+			fileList.FileInfos = map[string]*model.FileInfo{}
+			fileList.Order = []string{}
+			return nil
+		}
+		startInaccessibleIndex, endInaccessibleIndex := bounds.getInaccessibleRange(len(fileList.Order))
+		startInaccessibleCreatedAt := fileList.FileInfos[fileList.Order[startInaccessibleIndex]].CreateAt
+		endInaccessibleCreatedAt := fileList.FileInfos[fileList.Order[endInaccessibleIndex]].CreateAt
+		fileList.FirstInaccessibleFileTime = max(startInaccessibleCreatedAt, endInaccessibleCreatedAt)
+
+		files := fileList.FileInfos
+		order := fileList.Order
+		accessibleCount := bounds.end - bounds.start + 1
+		inaccessibleCount := lenFiles - accessibleCount
+		// Linearly cover shorter route to traverse files map
+		if inaccessibleCount < accessibleCount {
+			for i := 0; i < bounds.start; i++ {
+				delete(files, order[i])
+			}
+			for i := bounds.end + 1; i < lenFiles; i++ {
+				delete(files, order[i])
+			}
+		} else {
+			accessibleFiles := make(map[string]*model.FileInfo, accessibleCount)
+			for i := bounds.start; i <= bounds.end; i++ {
+				accessibleFiles[order[i]] = files[order[i]]
+			}
+			fileList.FileInfos = accessibleFiles
+		}
+
+		fileList.Order = fileList.Order[bounds.start : bounds.end+1]
+	} else {
+		linearFilterFileList(fileList, lastAccessibleFileTime)
+	}
+
+	return nil
+}
+
+// isInaccessibleFile indicates if the file is past the cloud plan's limit.
+func (a *App) isInaccessibleFile(file *model.FileInfo) (int64, *model.AppError) {
+	if file == nil {
+		return 0, nil
+	}
+
+	fl := &model.FileInfoList{
+		Order:     []string{file.Id},
+		FileInfos: map[string]*model.FileInfo{file.Id: file},
+	}
+
+	return fl.FirstInaccessibleFileTime, a.filterInaccessibleFiles(fl, filterFileOptions{assumeSortedCreatedAt: true})
```

Comment:
I think this might be risky to do. It seems the order of evaluation of `fl.FirstInaccessibleFileTime` and `a.filterInaccessibleFiles(fl, filterFileOptions{assumeSortedCreatedAt: true})` (which mutates `fl`) is unspecified according to the Go spec. 

See: https://go.dev/ref/spec#Order_of_evaluation

Relevant section reproduced below:

```
a := 1
f := func() int { a++; return a }
x := []int{a, f()}            // x may be [1, 2] or [2, 2]: evaluation order between a and f() is not specified
```

If I'm interpreting the spec correctly, this return is unsafe. The fix for this ambiguity is quite simple, we just split the return into two lines.  

```
err := a.filterInaccessibleFiles(fl, filterFileOptions{assumeSortedCreatedAt: true})
return fl.FirstInaccessibleFileTime, err
```

Reply:
> when evaluating the operands of an expression, assignment, or **return statement**, all function calls, method calls, and communication operations are evaluated in lexical left-to-right order.
> The expression list in the "return" statement may be a single call to a multi-valued function. The effect is as if each value returned from that function were assigned to a temporary variable with the type of the respective value, followed by a "return" statement listing these variables.

First point refers that it is safe in case of return statements.
And the 2nd point suggests that it does the same internally, what you have suggested.

> order of those events compared to the evaluation and indexing of x and the evaluation of y is not specified

It seems the order is unspecified when working with indexes.

But still for clarity, I'll update the code to manually assign and return.

Reply:
>But still for clarity, I'll update the code to manually assign and return.

SGTM